placeholders/placeholders.py

```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python3
+"""An encoder/decoder for placeholders in python3, using spm vocabulary"""
+import sys
+from typing import List, Dict, Type, Tuple, Optional
+from copy import deepcopy
+from dataclasses import dataclass
+import re
+import argparse
+from random import Random
+from sentencepiece import SentencePieceProcessor
+import yaml
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-c', '--config', type=str, help='Path to yaml configuration file, required for encoding')
+parser.add_argument('-m', '--mappings_file', type=str, default="mappings.yml", help='Path to the mappings, one yaml entry per line.')
+parser.add_argument('-s', '--seed', type=int, default=None, help='Seed for random number generator. Use `-` for sequential numbering.')
+mutex_group_1 = parser.add_mutually_exclusive_group(required=True)
+mutex_group_1.add_argument('--decode', action='store_true')
+mutex_group_1.add_argument('--encode', action='store_true')
+mutex_group_1.add_argument('--dump_placeholders', action='store_true', help='Check to print placeholders out')
+
+@dataclass
+class Rule:
+    """Just a wrapper for regex rules"""
+    pattern: str
+
+@dataclass
+class Configuration:
+    """Object holding the yaml config"""
+    def __init__(self, config_file):
+        with open(config_file, 'r') as config_handle:
+            my_config = yaml.safe_load(config_handle)
+
+        # Parse
+        self.rules = [Rule(regex) for regex in my_config['regexes']]
+        self.placeholder_symbol = my_config.get('placeholder-symbol', '@')
+        self.num_placeholders = my_config.get('num-placeholders', 20)
+        self.placeholders = [self.placeholder_symbol + str(i) for i in range(self.num_placeholders)]
+
+        # Add a rule that escapes patterns that look like a placeholder already
+        # TODO: this will match placeholders we can't reach because `num_placeholders` might be smaller
+        # causing us to replace placeholders that will never be produced by us, and draining our
+        # available placeholders.
+        self.rules.append(Rule(pattern=re.escape(self.placeholder_symbol) + r'\d+'))
```

Comment:
This is the edge case that still breaks stuff. The pattern added will match things like `PLACEHOLDER100` even when `num_placeholders=4`. Ideally we would have the regex only match `[0-4]` in that case but generating a regex that matches a range [is hard](https://stackoverflow.com/questions/19269448/regular-expression-number-range).

Reply:
We can pre-generate all ever available placeholders and then match those explicitly? But this will only happen if our placeholder token is already in the input text which  we assume wouldn't happen, correct?

Reply:
Correct